bool[4] values
bool[4] mask